source/gui/addonGui.py

```diff
@@ -270,141 +270,10 @@ def onAddClick(self, evt):
-					_("Error"),
-					wx.OK | wx.ICON_ERROR
-				)
-				return
-
-			if not addonVersionCheck.hasAddonGotRequiredSupport(bundle):
-				self._showAddonRequiresNVDAUpdateDialog(bundle)
-				return
-			elif not addonVersionCheck.isAddonTested(bundle):
-				self._showAddonUntestedDialog(bundle)
-				return
-			elif wx.YES != self._showConfirmAddonInstallDialog(bundle):
-				return
-
-			prevAddon=None
-			for addon in self.curAddons:
-				if not addon.isPendingRemove and bundle.name==addon.manifest['name']:
-					prevAddon=addon
-					break
-			if prevAddon:
-				summary=bundle.manifest["summary"]
-				curVersion=prevAddon.manifest["version"]
-				newVersion=bundle.manifest["version"]
-				if gui.messageBox(
-					# Translators: A message asking if the user wishes to update an add-on with the same version currently installed according to the version number.
-					_("You are about to install version {newVersion} of {summary}, which appears to be already installed. Would you still like to update?").format(
-						summary=summary,
-						newVersion=newVersion
-					)
-					if curVersion==newVersion else
-					# Translators: A message asking if the user wishes to update a previously installed add-on with this one.
-					_("A version of this add-on is already installed. Would you like to update {summary} version {curVersion} to version {newVersion}?").format(
-						summary=summary,
-						curVersion=curVersion,
-						newVersion=newVersion
-					),
-					# Translators: A title for the dialog  asking if the user wishes to update a previously installed add-on with this one.
-					_("Add-on Installation"),
-					wx.YES|wx.NO|wx.ICON_WARNING)!=wx.YES:
-						return
-				prevAddon.requestRemove()
-			progressDialog = gui.IndeterminateProgressDialog(gui.mainFrame,
-			# Translators: The title of the dialog presented while an Addon is being installed.
-			_("Installing Add-on"),
-			# Translators: The message displayed while an addon is being installed.
-			_("Please wait while the add-on is being installed."))
-			try:
-				gui.ExecAndPump(addonHandler.installAddonBundle,bundle)
-			except:
-				log.error("Error installing  addon bundle from %s"%addonPath,exc_info=True)
-				self.refreshAddonsList()
-				progressDialog.done()
-				del progressDialog
-				# Translators: The message displayed when an error occurs when installing an add-on package.
-				gui.messageBox(_("Failed to install add-on  from %s")%addonPath,
-					# Translators: The title of a dialog presented when an error occurs.
-					_("Error"),
-					wx.OK | wx.ICON_ERROR)
-				return
-			else:
-				self.refreshAddonsList(activeIndex=-1)
-				progressDialog.done()
-				del progressDialog
-		finally:
-			if closeAfter:
-				# #4460: If we do this immediately, wx seems to drop the WM_QUIT sent if the user chooses to restart.
-				# This seems to have something to do with the wx.ProgressDialog.
-				# The CallLater seems to work around this.
-				wx.CallLater(1, self.Close)

```

Comment:
I guess the context manager is the alternative approach for this? Are you sure that it won't introduce #4460 again?

Reply:
Quite sure, this `self.Close` is for the `AddonManger` dialog. The call to `wx.CallLater` only happened when an instance of the `AddonManager` had to be created for the purpose of installing an addon from the shell integration. When there was an `AddonManager` dialog was already opened, or installation was triggered by the `AddonManager` then this wasn't hit. I expect part of the problem was in calling `del progressDialog` rather than `Destroy` as well. But either way, it's no longer necessary to destroy the `progressDialog` and the `AddonManger` at once.